src/prop/prop_proof_manager.cpp

```diff
@@ -194,5 +229,163 @@ std::shared_ptr<ProofNode> PropPfManager::getProof(bool connectCnf)
+  }
+  return pfs;
+}
+
+void PropPfManager::notifyExplainedPropagation(TrustNode trn)
+{
+  Node proven = trn.getProven();
+  // If we are not producing proofs in the theory engine there is no need to
+  // keep track in d_proof of the clausification. We still need however to let
+  // the SAT proof manager know that this clause is an assumption.
+  bool proofLogging = trn.getGenerator() != nullptr;
+  Trace("cnf")
+      << "PropPfManager::notifyExplainedPropagation: proven explanation"
+      << proven << ", proofLogging=" << proofLogging << "\n";
+  if (proofLogging)
+  {
+    Assert(trn.getGenerator()->getProofFor(proven)->isClosed());
+    Trace("cnf-steps") << proven << " by explainPropagation "
+                       << trn.identifyGenerator() << std::endl;
+    d_proof.addLazyStep(proven,
+                        trn.getGenerator(),
+                        TrustId::NONE,
+                        true,
+                        "PropPfManager::notifyExplainedPropagation");
+  }
+  // since the propagation is added directly to the SAT solver via theoryProxy,
+  // do the transformation of the lemma E1 ^ ... ^ En => P into CNF here
+  NodeManager* nm = NodeManager::currentNM();
+  Node clauseImpliesElim;
+  if (proofLogging)
+  {
+    clauseImpliesElim = nm->mkNode(Kind::OR, proven[0].notNode(), proven[1]);
+    Trace("cnf") << "PropPfManager::notifyExplainedPropagation: adding "
+                 << ProofRule::IMPLIES_ELIM << " rule to conclude "
+                 << clauseImpliesElim << "\n";
+    d_proof.addStep(clauseImpliesElim, ProofRule::IMPLIES_ELIM, {proven}, {});
+  }
+  Node clauseExp;
+  // need to eliminate AND
+  if (proven[0].getKind() == Kind::AND)
+  {
+    std::vector<Node> disjunctsAndNeg{proven[0]};
+    std::vector<Node> disjunctsRes;
+    for (unsigned i = 0, size = proven[0].getNumChildren(); i < size; ++i)
+    {
+      disjunctsAndNeg.push_back(proven[0][i].notNode());
+      disjunctsRes.push_back(proven[0][i].notNode());
+    }
+    disjunctsRes.push_back(proven[1]);
+    clauseExp = nm->mkNode(Kind::OR, disjunctsRes);
+    if (proofLogging)
+    {
+      // add proof steps to convert into clause
+      Node clauseAndNeg = nm->mkNode(Kind::OR, disjunctsAndNeg);
+      d_proof.addStep(clauseAndNeg, ProofRule::CNF_AND_NEG, {}, {proven[0]});
+      d_proof.addStep(clauseExp,
+                      ProofRule::RESOLUTION,
+                      {clauseAndNeg, clauseImpliesElim},
+                      {nm->mkConst(true), proven[0]});
+    }
+  }
+  else
+  {
+    clauseExp = nm->mkNode(Kind::OR, proven[0].notNode(), proven[1]);
+  }
+  d_currPropagationProcessed = normalizeAndRegister(clauseExp, false);
+  // consume steps if clausification being recorded. If we are not logging it,
+  // we need to add the clause as a closed step to the proof so that the SAT
+  // proof does not have non-input formulas as assumptions. That clause is the
+  // result of normalizeAndRegister, stored in d_currPropagationProcessed
```

Comment:
```suggestion
  // If we are not logging the clausification, we need to add the clause, as *it
  // will be saved in the SAT solver* (i.e., as clauseExp), as closed step in
  // the d_proof, so that there are no non-input assumptions.
  ```